hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -1056,8 +1056,18 @@ private TcpClientConnection onAuthenticated(TcpClientConnection connection,
                     });
                 } else {
                     establishedInitialClusterConnection = true;
-                    clientState = ClientState.INITIALIZED_ON_CLUSTER;
-                    fireLifecycleEvent(LifecycleState.CLIENT_CONNECTED);
+                    if (!asyncStart) {
+                        // For a sync start client, we will send the client state as the last statement of Client instance
+                        // construction. Therefore, we can change client state and send the event here.
+                        clientState = ClientState.INITIALIZED_ON_CLUSTER;
+                        fireLifecycleEvent(LifecycleState.CLIENT_CONNECTED);
+                    } else {
+                        // For async clients, we return the client instance to the user without waiting for the client
+                        // state to be sent. The state should be INITIALIZED_ON_CLUSTER after we send the client state.
+                        // Also the CLIENT_CONNECTED event should be fired after the state is sent. initializeClientOnCluster
+                        // will handle all of that.
+                        executor.execute(() -> initializeClientOnCluster(clusterId));
```

Comment:
we can actually. Let me see

Reply:
In order to make the client throw exceptions, I think we should stay in the initial state. If client state is CONNECTED_TO_CLUSTER, it will throw `IOException` in checkInvocationAllowed method and it will lead to retrying the invocation for invocation timeout duration. I think we want to throw HazelcastClientOfflineException instead.